### PR TITLE
nixos: systemd-homed support

### LINKED ISSUE #91243: systemd-homed support
**Project description**
One of many systemd utilities that improve home directories management
**Metadata**
https://wiki.archlinux.org/index.php/Systemd-homed
https://systemd.io/HOME_DIRECTORY/

* homepage URL: https://systemd.io/HOME_DIRECTORY/
* source URL: https://github.com/systemd/systemd
* license: GPL-2.0
* platforms:  linux

### PR DESCRIPTION
**Pull request by alaviss**
###### Description of changes

Fixes #91243 #79263

This change involves two new modules, one for `userdbd` and another for `homed`.

I've been daily driving this for awhile, and I can verify that it works with `fprintd` as well.

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by flokli** [comment]
Thanks for the PR! However, this needs to target staging due to the amount of rebuilds. Can this be rebased on staging? (Please check the contribution guideline on how to rebase this on staging without causing a mass ping)

File: nixos/tests/systemd-userdbd.nix [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+    import json
+    from shlex import quote
+
+    def getUserRecord(name):
+      Interface = "unix:/run/systemd/userdb/io.systemd.Multiplexer/io.systemd.UserDatabase"
+      payload = json.dumps({
+        "service": "io.systemd.Multiplexer",
+        "userName": name
+      })
+      return json.loads(machine.succeed(f"varlink call {Interface}.GetUserRecord {quote(payload)}"))
```
Comment (flokli) [inline nixos/tests/systemd-userdbd.nix:26]: Can you add some tests using`getent` here, to make sure the NSS side of things also works?

Or could this potentially be folded into `nixos/tests/nscd.nix`, having one machine config with and one without userdbd enabled?

Comment (flokli) [reply nixos/tests/systemd-userdbd.nix:26]: Hmmh, we could hide `/etc/userdb/test-user-dropin.user` for the getent process with libredirect, like we hide `/etc/hosts` (so these lookups can only succeed if they go via ns(n)cd).

But it's a bit ugly indeed, and the current `whoami` check is probably fine to ensure it gets the user id somehow?! WDYT?

Comment (alaviss) [reply nixos/tests/systemd-userdbd.nix:26]: > Hmmh, we could hide `/etc/userdb/test-user-dropin.user` for the getent process with libredirect, like we hide `/etc/hosts` (so these lookups can only succeed if they go via ns(n)cd).

Yea but all lookup using `nss-systemd` has to go through `nscd` which can see all files anyway :p

I think for userdbd this simple check is enough.

**Comment by flokli** [comment]
Alright, let's cook this in staging :-) Thanks for the PR! :-)

**Pull request merged**